README.md
libyui-mga-qt
=============

libyui-qt plugins/extensions for the Mageia tools


ChangeLog
-------------------------------------------------------------------
Sun Jul 11 16:09:45 CEST 2021 - anaselli@linux.it
- New YMGA_CBTable implementation
- built against new libyui tree.
- 1.2.0

-------------------------------------------------------------------
Mon Jun 29 19:19:01 CEST 2020 - anaselli@linux.it
- Added menubar and new menu items implementation
- 1.1.0

-------------------------------------------------------------------
Sun Jan 12 15:09:45 CET 2020 - anaselli@linux.it
- fix issue 1 Using spacebar to uncheck packages does not mark them for deletion
- 1.0.6

-------------------------------------------------------------------
Sun Dec 16 21:21:42 CET 2018 - anaselli@linux.it
- fix build against qt-5.11.0
- 1.0.5

-------------------------------------------------------------------
Sun Jan 28 17:38:48 CET 2018 - anaselli@linux.it
- tagged version
- 1.0.4

-------------------------------------------------------------------
Mon Oct 19 23:51:42 CEST 2015 - anaselli@linux.it
- new libyui major so version
- 1.0.3

-------------------------------------------------------------------
Sun Jan 19 13:14:40 CET 2014 anaselli@linux.it

- Ported to Qt5 (aligned to yui-qt)
- 1.0.2



src/YQWE.h
/*
  Copyright (C) 2000-2012 Novell, Inc
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YQWE.h

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#ifndef YQWE_h
#define YQWE_h
#include <YExternalWidgets.h>

class YQWE : public YExternalWidgets
{
public:
  YQWE( const std::string& name );
  virtual ~YQWE();

protected:

  virtual YExternalWidgetFactory * createExternalWidgetFactory();
};

#endif //YQWE_h

src/YQWE.cc
/*
  Copyright (C) 2000-2012 Novell, Inc
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YQWE.cc

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#include "YQWE.h"
#include "YMGAQWidgetFactory.h"
#include <YUIException.h>
#include <string.h>

YQWE::YQWE( const std::string& name ) : YExternalWidgets(name)
{
}

YQWE::~YQWE()
{
}

YExternalWidgetFactory* YQWE::createExternalWidgetFactory()
{
  YMGAQWidgetFactory * factory = new YMGAQWidgetFactory();
  YUI_CHECK_NEW( factory );

  return factory;
}

YExternalWidgets * createExternalWidgets(const char *name)
{
  std::string pluginName(name);
  YQWE * ew = new YQWE( pluginName );
  YUI_CHECK_NEW ( ew );

  return YExternalWidgets::externalWidgets(pluginName);
}


src/YMGAQMenuBar.h
/*
  Copyright 2020 by Angelo Naselli <anaselli at linux dot it>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YMGAQMenuBar.h

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/


#ifndef YMGAQMenuBar_h
#define YMGAQMenuBar_h
#include <QWidget>
#include <QMenu>

#include <yui/mga/YMGAMenuBar.h>


class YMGAQMenuBar : public QWidget, public YMGAMenuBar
{
    Q_OBJECT

public:

    /**
     * Constructor.
     **/
    YMGAQMenuBar( YWidget * parent );

    /**
     * Destructor.
     **/
    virtual ~YMGAQMenuBar();

    /**
     * Add an YMenuItem first item represents the menu name, other sub items menu entries
     *
     * Reimplemented from YSelectionWidget.
     **/
    virtual void addItem( YItem * item );

    /**
     * Preferred width of the widget.
     *
     * Reimplemented from YWidget.
     **/
    virtual int preferredWidth();

    /**
     * Preferred height of the widget.
     *
     * Reimplemented from YWidget.
     **/
    virtual int preferredHeight();

    /**
     * Set the new size of the widget.
     *
     * Reimplemented from YWidget.
     **/
    virtual void setSize( int newWidth, int newHeight );

    /**
    * Enable YMGAMenuItem (menu name or menu entry) to enable/disable it into menubar or menu
    *
    * Reimplemented from YMGAMenuBar.
    **/
    virtual void enableItem(YItem * menu_item, bool enable=true);

    /**
    * Hide YMGAMenuItem (menu name or menu entry) to hide/show it into menubar or menu
    *
    * Reimplemented from YMGAMenuBar.
    **/
    virtual void hideItem(YItem * menu_item, bool invisible=true);

    /**
     * Delete all items.
     *
     * Reimplemented from YSelectionWidget
     **/
    virtual void deleteAllItems();


protected:

    /**
     * Add a submenu
     *
     **/
    virtual void addSubMenu( QMenu * menu, YItem * item );

    /**
     * Add an action (a menu entry without submenus)
     *
     **/
    virtual void addAction( QMenu * menu, YItem * item );

private:
    struct Private;
    Private *d;

};

#endif //YMGAQMenuBar_h


src/YMGAQMenuBar.cc
/*
  Copyright 2020 by Angelo Naselli <anaselli at linux dot it>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YMGAQMenuBar.cc

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#include <QMenu>
#include <QMenuBar>
#include <QHBoxLayout>
#include <QAction>
#include <QFileInfo>
#define YUILogComponent "mga-qt-ui"
#include <yui/YUILog.h>

#include <yui/qt/utf8.h>

#include <yui/qt/YQUI.h>
#include <yui/YEvent.h>
#include <yui/qt/YQSignalBlocker.h>
#include <yui/YUIException.h>
#include <yui/YMenuItem.h>

#include "YMGAQMenuBar.h"
#include <yui/qt/YQApplication.h>
#include <yui/mga/YMGAMenuItem.h>

typedef std::map<YItem*, QObject*> MenuEntryMap;
typedef std::pair<YItem*, QObject*> MenuEntryPair;

struct YMGAQMenuBar::Private
{
  QMenuBar *menubar;
  MenuEntryMap menu_entry;
};


class QItemAction : public QAction
{

private:
  YMenuItem *_yitem;
public:
  QItemAction(YMenuItem *item, QObject *parent = nullptr ) :
    QAction(item->label().c_str(), parent), _yitem(item)
  {

  }

  virtual ~QItemAction() {}

  void selectedItem()
  {
    YQUI::ui()->sendEvent( new YMenuEvent( _yitem ) );
  }
};

YMGAQMenuBar::YMGAQMenuBar( YWidget * parent )
    : QWidget( (QWidget *) parent->widgetRep() )
    , YMGAMenuBar( parent ), d(new Private)
{
  YUI_CHECK_NEW ( d );

  setWidgetRep ( this );

  QHBoxLayout* layout = new QHBoxLayout ( this );
  layout->setSpacing ( 0 );
  setLayout ( layout );

  layout->setMargin ( YQWidgetMargin );

  d->menubar = new QMenuBar ( this );
  YUI_CHECK_NEW ( d->menubar );
  layout->addWidget ( d->menubar );
}


YMGAQMenuBar::~YMGAQMenuBar()
{
      delete d;
}

void YMGAQMenuBar::addAction(QMenu* menu, YItem* yitem)
{
  YMenuItem * item = dynamic_cast<YMenuItem *> ( yitem );
  YUI_CHECK_PTR ( item );

  YMenuSeparator *separator = dynamic_cast<YMenuSeparator *>(yitem);
  if (separator)
    menu->addSeparator();
  else
  {
    // TODO icon from item
    QItemAction *action = new QItemAction(item, this);
    if (item->hasIconName())
    {
      QString qtIcon = fromUTF8( yitem->iconName() );
      QString icon_name = QFileInfo( qtIcon ).baseName();

      if ( QIcon::hasThemeIcon( icon_name ) )
      {
        action->setIcon( QIcon::fromTheme ( icon_name ) );
      }
      else
      {
          QPixmap pixmap( qtIcon );

          if ( !pixmap.isNull() )
              action->setIcon( QIcon( pixmap ) );
      }
      action->setIconVisibleInMenu(true);
    }
    connect(action, &QAction::triggered, action, &QItemAction::selectedItem);
    menu->addAction(action);

    YMGAMenuItem *menuItem = dynamic_cast<YMGAMenuItem *>(yitem);
    if (menuItem)
    {
      action->setEnabled(menuItem->enabled());
      d->menu_entry.insert(MenuEntryPair(yitem, action));
      action->setVisible(!menuItem->hidden());
    }
  }

}

void YMGAQMenuBar::addSubMenu(QMenu* menu, YItem* yitem)
{
  YMenuItem * item = dynamic_cast<YMenuItem *> ( yitem );
  YUI_CHECK_PTR ( item );

  // TODO icon from item
  QMenu *m=menu->addMenu(item->label().c_str());
  if (item->hasChildren())
  {
    YMGAMenuItem *menuItem = dynamic_cast<YMGAMenuItem *>(yitem);
    if (menuItem)
    {
      m->setEnabled(menuItem->enabled());
      d->menu_entry.insert(MenuEntryPair(yitem, m));
      m->menuAction()->setVisible(!menuItem->hidden());
    }
    for (YItemIterator miter = item->childrenBegin(); miter != item->childrenEnd(); miter++)
    {
      YMenuItem *m_item= dynamic_cast<YMenuItem *>(*miter);
      if (m_item->hasChildren())
      {
        addSubMenu(menu, m_item);
      }
      else
      {
        addAction(m, m_item);
      }
    }
  }
}


void YMGAQMenuBar::addItem(YItem* yitem)
{
  YMenuItem * item = dynamic_cast<YMenuItem *> ( yitem );
  YUI_CHECK_PTR ( item );

  // TODO icon from item
  QMenu *menu = d->menubar->addMenu(item->label().c_str());

  if (item->hasChildren())
  {
    for (YItemIterator miter = item->childrenBegin(); miter != item->childrenEnd(); miter++)
    {
      YMenuItem *m_item= dynamic_cast<YMenuItem *>(*miter);
      if (m_item->hasChildren())
      {
        addSubMenu(menu, m_item);
      }
      else
      {
        addAction(menu, m_item);
      }
    }
  }

  YMGAMenuItem *menuItem = dynamic_cast<YMGAMenuItem *>(yitem);
  if (menuItem)
  {
    menu->setEnabled(menuItem->enabled());
    d->menu_entry.insert(MenuEntryPair(yitem, menu));
    menu->menuAction()->setVisible(!menuItem->hidden());
  }
  YMGAMenuBar::addItem(yitem);
  menu->update();
}

int YMGAQMenuBar::preferredWidth()
{
    // Arbitrary value.
    // Use a MinSize widget to set a size that is useful for the application.

    return 20 + d->menubar->sizeHint().width();
}


int YMGAQMenuBar::preferredHeight()
{
    // Arbitrary value.
    // Use a MinSize widget to set a size that is useful for the application.
    return 20 + d->menubar->sizeHint().height();
}



void YMGAQMenuBar::setSize( int newWidth, int newHeight )
{
    resize( newWidth, newHeight );
}

void YMGAQMenuBar::enableItem(YItem* menu_item, bool enable)
{
  YMGAMenuBar::enableItem(menu_item, enable);

  auto search = d->menu_entry.find( menu_item );
  if (search != d->menu_entry.end())
  {
    QMenu * menu_entry = dynamic_cast<QMenu*>(search->second);
    if (menu_entry)
      menu_entry->setEnabled(enable);
    else
    {
      QAction * menu_action = dynamic_cast<QAction*>(search->second);
      if (menu_action)
        menu_action->setEnabled(enable);
    }
  }
  else
  {
    yuiError() << menu_item->label() << " not found" << std::endl;
  }
}

void YMGAQMenuBar::hideItem(YItem* menu_item, bool invisible)
{
  YMGAMenuBar::hideItem(menu_item, invisible);

  auto search = d->menu_entry.find( menu_item );
  if (search != d->menu_entry.end())
  {
    QMenu * menu_entry = dynamic_cast<QMenu*>(search->second);
    if (menu_entry)
    {
      menu_entry->menuAction()->setVisible(!invisible);

//       resize( d->menubar->sizeHint() + QSize(20,20));
//       update();

    }
    else
    {
      QAction * menu_action = dynamic_cast<QAction*>(search->second);
      if (menu_action)
        menu_action->setVisible(!invisible);
    }
  }
  else
  {
    yuiError() << menu_item->label() << " not found" << std::endl;
  }
}

void YMGAQMenuBar::deleteAllItems()
{
  d->menubar->clear();
//   for (MenuEntryMap::iterator it=d->menu_entry.begin(); it!=d->menu_entry.end(); ++it)
//   {
//     QMenu * menu_entry = dynamic_cast<QMenu*>(it->second);
//     if (menu_entry)
//     {
//       d->menubar->removeAction(menu_entry->menuAction());
//     }
//     else
//     {
//       QAction * menu_action = dynamic_cast<QAction*>(it->second);
//       if (menu_action)
//           d->menubar->removeAction(menu_action);
//     }
//   }

  d->menu_entry.clear();

  YSelectionWidget::deleteAllItems();
}




src/YMGA_QCBTable.h
/*
  Copyright 2013-2020 by Angelo Naselli <anaselli at linux dot it>

  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YMGA_QCBTable.h

  Author:       Angelo Naselli <anaselli@linux.it>

/-*/

#ifndef YMGA_QCBTable_h
#define YMGA_QCBTable_h
#include <QFrame>

#include <yui/mga/YMGA_CBTable.h>
#include <yui/qt/QY2ListView.h>


// class QY2ListView;
class QTreeWidgetItem;
class YQListViewItem;
class YMGA_QCBTableListViewItem;

class YMGA_QCBTable : public QFrame, public YMGA_CBTable
{
  Q_OBJECT

public:

  /**
   * Constructor.
   **/
    YMGA_QCBTable(YWidget* parent, YCBTableHeader* tableHeader);

  /**
   * Destructor.
   **/
  virtual ~YMGA_QCBTable();

  /**
   * Switch between sorting by item insertion order (keepSorting: true) or
   * allowing the user to sort by an arbitrary column (by clicking on the
   * column header).
   *
   * Reimplemented from YTable.
   **/
  virtual void setKeepSorting( bool keepSorting );

  /**
   * Add an item.
   *
   * Reimplemented from YSelectionWidget.
   **/
  virtual void addItem( YItem * item );

  /**
   * Add multiple items.
   *
   * Reimplemented for efficiency from YSelectionWidget.
   **/
  virtual void addItems( const YItemCollection & itemCollection );

  /**
   * Select or deselect an item.
   *
   * Reimplemented from YSelectionWidget.
   **/
  virtual void selectItem( YItem * item, bool selected = true );

  /**
   * Deselect all items.
   *
   * Reimplemented from YSelectionWidget.
   **/
  virtual void deselectAllItems();

  /**
   * Delete all items.
   *
   * Reimplemented from YSelectionWidget.
   **/
  virtual void deleteAllItems();

  /**
   * Notification that a cell (its text and/or its icon) was changed from the
   * outside. Applications are required to call this whenever a table cell is
   * changed after adding the corresponding table item (the row) to the table
   * widget.
   *
   * Reimplemented from YTable.
   **/
  virtual void cellChanged( const YTableCell * cell );

  /**
   * Set enabled/disabled state.
   *
   * Reimplemented from YWidget.
   **/
  virtual void setEnabled( bool enabled );

  /**
   * Preferred width of the widget.
   *
   * Reimplemented from YWidget.
   **/
  virtual int preferredWidth();

  /**
   * Preferred height of the widget.
   *
   * Reimplemented from YWidget.
   **/
  virtual int preferredHeight();

  /**
   * Set the new size of the widget.
   *
   * Reimplemented from YWidget.
   **/
  virtual void setSize( int newWidth, int newHeight );

  /**
   * Accept the keyboard focus.
   *
   * Reimplemented from YWidget.
   **/
  virtual bool setKeyboardFocus();

  /**
   * check/uncheck Item from application.
   *
   * Reimplemented from YMGA_CBTable
   * Note that setting it from cell with setChecked(true) does not update
   * the table
   *
   **/
  virtual void setItemChecked(YItem* yitem, int column, bool checked = true);


protected slots:

  /**
   * Notification that an item is selected (single click or keyboard).
   **/
  void slotSelected( QTreeWidgetItem * );

  /**
   * Notification that the item selection changed
   * (relevant for multiSelection mode).
   **/
  // YMGA_CBTable does not work in multiSelection Mode
  //void slotSelectionChanged();

  /**
   * Notification that an item is activated (double click or keyboard).
   **/
  void slotActivated( QTreeWidgetItem * );

  /**
   * Propagate an "item expanded" event to the underlying YTableItem.
   **/
  void slotItemExpanded( QTreeWidgetItem * item );

  /**
   * Propagate an "item collapsed" event to the underlying YTableItem.
   **/
  void slotItemCollapsed( QTreeWidgetItem * item );


  /**
   * Propagate a context menu selection.
   *
   * This will trigger a 'ContextMenuActivated' event if 'notifyContextMenu' is set.
   **/
  void slotContextMenu ( const QPoint & pos );

  /**
   * Propagate an "item changed" or an "item clicked" event to the underlying YTableItem..
   *
   **/
  void slotColumnClicked(QTreeWidgetItem *     item,
                         int                   col);


protected:

  /**
   * Select the original item (the YTableItem) that corresponds to the
   * specified listViewItem.
   **/
  void selectOrigItem( QTreeWidgetItem * listViewItem );

  /**
   * Internal addItem() method that will not do expensive operations in batch
   * mode.
   **/
  void addItem( YItem * item, bool batchMode, bool resizeColumnsToContent );

  /**
   * Clone (create Qt item counterparts) for all child items of 'parentItem'.
   * Set their Qt item parent to 'parentItemClone'.
   **/
  void cloneChildItems( YCBTableItem              * parentItem,
                        YMGA_QCBTableListViewItem * parentItemClone );

  //
  // Data members
  //

  QY2ListView * _qt_listView;

};



/**
 * Visual representation of a YCBTableItem.
 **/
class YMGA_QCBTableListViewItem: public QY2ListViewItem
{
public:

  /**
   * Constructor for toplevel items.
   **/
  YMGA_QCBTableListViewItem( YMGA_QCBTable * table,
                             QY2ListView   * parent,
                             YCBTableItem  * origItem );

  /**
   * Constructor for nested items.
   **/
  YMGA_QCBTableListViewItem( YMGA_QCBTable             * table,
                             YMGA_QCBTableListViewItem * parentItemClone,
                             YCBTableItem              * origItem );


  /**
   * Return the parent table widget.
   **/
  YMGA_QCBTable * table() const { return _table; }

  /**
   * Return the corresponding YCBTableItem.
   **/
  YCBTableItem * origItem() const { return _origItem; }

  /**
   * Update this item's display with the content of 'cell'.
   **/
  void updateCell( const YCBTableCell * cell );

  /**
   * Update all columns of this item with the content of the original item.
   **/
  void updateCells();

  /**
   * The text of the table cell or the sort-key if available.
   **/
  virtual QString smartSortKey(int column) const override;

protected:
  /**
   * Common initializations for all constructors
   **/
  void init();

  /**
   * Set the alignment for each column according to the YTable parent's
   * alignment.
   **/
  void setColAlignment();

  YMGA_QCBTable * _table;
  YCBTableItem  * _origItem;
};



#endif // YQLabel_h


src/YMGAQWidgetFactory.h
/*
  Copyright (C) 2000-2012 Novell, Inc
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:		YMGAQWidgetFactory.h

  Author:	Stefan Hundhammer <sh@suse.de>

/-*/

#ifndef YMGAQWidgetFactory_h
#define YMGAQWidgetFactory_h


#include <yui/mga/YMGAWidgetExtensionFactory.h>

#include "YMGA_CBTable.h"
#include "YMGAMenuBar.h"

using std::string;


/**
 * Concrete widget factory for mandatory widgets.
 **/
class YMGAQWidgetFactory: public YMGAWidgetFactory
{
public:

  virtual YMGA_CBTable * createCBTable ( YWidget * parent, YTableHeader * header_disown );

  virtual YMGAMenuBar * createMenuBar ( YWidget * parent );

protected:

    friend class YQWE;

    /**
     * Constructor.
     *
     * Use YUI::widgetFactory() to get the singleton for this class.
     **/
    YMGAQWidgetFactory();

    /**
     * Destructor.
     **/
    virtual ~YMGAQWidgetFactory();

}; // class YWidgetFactory


#endif // YMGAQWidgetFactory_h


src/YMGAQWidgetFactory.cc
/*
  Copyright (C) 2000-2012 Novell, Inc
  This library is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as
  published by the Free Software Foundation; either version 2.1 of the
  License, or (at your option) version 3.0 of the License. This library
  is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or
  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
  License for more details. You should have received a copy of the GNU
  Lesser General Public License along with this library; if not, write
  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
  Floor, Boston, MA 02110-1301 USA
*/


/*-/

  File:         YMGAQWidgetFactory.cc

  Author:       Stefan Hundhammer <sh@suse.de>

/-*/

#include <QPoint>
#define YUILogComponent "mga-qt"
#include <yui/YUILog.h>

#include "YMGAQWidgetFactory.h"
#include <yui/qt/YQApplication.h>
#include <yui/qt/YQUI.h>
#include <yui/YUIException.h>
#include <YExternalWidgets.h>


#include <string>

#include "YMGA_QCBTable.h"
#include "YMGAQMenuBar.h"

using std::string;


YMGAQWidgetFactory::YMGAQWidgetFactory()
    : YMGAWidgetFactory()
{
    // NOP
}


YMGAQWidgetFactory::~YMGAQWidgetFactory()
{
    // NOP
}


YMGA_CBTable *
YMGAQWidgetFactory::createCBTable( YWidget * parent, YTableHeader * header )
{
    YCBTableHeader *hdr = dynamic_cast<YCBTableHeader *>(header);
    YUI_CHECK_NEW(hdr);
    YMGA_QCBTable * table = new YMGA_QCBTable( parent, hdr );
    YUI_CHECK_NEW( table );

    return table;
}

YMGAMenuBar * YMGAQWidgetFactory::createMenuBar(YWidget* parent)
{
    YMGAMenuBar *menubar = new YMGAQMenuBar( parent );
    YUI_CHECK_NEW( menubar );

    return menubar;
}


